src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -659,6 +667,114 @@ internal static bool IsModuleMatchingModuleSpec(out ModuleMatchFailure matchFail
+            // (the required path may still refer to some nested module though)
+            if (!modulePath.StartsWith(requiredPath, strcmp))
+            {
+                return false;
+            }
+
+            // Otherwise, the required module may point to the module base directory
+            string moduleDirPath = Path.GetDirectoryName(modulePath);
+
+            // The module itself may be in a versioned directory
+            if (Version.TryParse(Path.GetFileName(moduleDirPath), out Version unused))
+            {
+                moduleDirPath = Path.GetDirectoryName(moduleDirPath);
+            }
+
+            return moduleDirPath.Equals(requiredPath, strcmp);
+        }
+
+        /// <summary>
+        /// Takes the name of a module as used in a module specification
+        /// and either returns it as a simple name (if it was a simple name)
+        /// or a fully qualified, PowerShell-resolved path.
+        /// </summary>
+        /// <param name="moduleName">The name or path of the module from the specification.</param>
+        /// <param name="basePath">The path to base relative paths off.</param>
+        /// <param name="executionContext">The current execution context.</param>
+        /// <returns>
+        /// The simple module name if the given one was simple,
+        /// otherwise a fully resolved, absolute path to the module.
+        /// </returns>
+        /// <remarks>
+        /// 2018-11-09 rjmholt:
+        /// There are several, possibly inconsistent, path handling mechanisms
+        /// in the module cmdlets. After looking through all of them and seeing
+        /// they all make some assumptions about their caller I wrote this method.
+        /// Hopefully we can find a standard path resolution API to settle on.
+        /// </remarks>
+        internal static string NormalizeModuleName(
+            string moduleName,
+            string basePath,
+            ExecutionContext executionContext)
+        {
+            if (moduleName == null)
+            {
+                return null;
+            }
+
+            // Check whether the module is a path -- if not, it is a simple name and we just return it.
+            if (!IsModuleNamePath(moduleName))
+            {
+                return moduleName;
+            }
+
+            // Standardize directory separators
+            moduleName = moduleName.Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
+
+            // Path.IsRooted() is deprecated -- see https://github.com/dotnet/corefx/issues/22345
+            if (!Path.IsPathFullyQualified(moduleName))
+            {
+                moduleName = Path.Join(basePath, moduleName);
+            }
+
+            // Use the PowerShell filesystem provider to fully resolve the path
+            return ModuleCmdletBase.GetResolvedPath(moduleName, executionContext).TrimEnd(Path.DirectorySeparatorChar);
```

Comment:
What would happen if `GetResolvedPath` cannot resolve `moduleName`?

Reply:
We end up in this logic:
https://github.com/PowerShell/PowerShell/blob/caad7ed471d46629f155a2c7ccbbafa4ffc29355/src/System.Management.Automation/namespaces/LocationGlobber.cs#L121-L246

I think it's relatively safe and well tested -- it doesn't look like it's designed to throw.

The call to `ModuleBase.GetResolvedPath()` itself surrounds the globber call with a `try`/`catch` and if anything bad happens returns `null`.

I've added a line to check for the `null` return and replace it with the original "un-resolved" module name, just in case.

Reply:
Yes, null reference is my main concern here. As long as that's guarded, it should be good.